### PR TITLE
benchmarking block encoder

### PR DESCRIPTION
**Pull request by yunxing**
enchmarkCRCEncoding	2000000000	         0.26 ns/op	3827.48 MB/s
ok  	github.com/cfs/disk	64.053s

### CODE REVIEW

File: disk/block.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+//
+// If an EOF happends before there are less than b.size - 4
+// bytes were read, the crc32 is calculated on the partial
+// bytes. However, there will still be a crc32 + b.size - 4
+// bytes written to the underlying writer, the missing bytes
+// will be filled with 0.
+func (b *blockEncoder) encodeOne(r io.Reader) (int64, error) {
+	payload := b.buf[4:]
+	n, err := io.ReadAtLeast(r, payload, int(b.size - 4))
+	if err != nil && err != io.ErrUnexpectedEOF {
```
Comment (xiang90) [inline disk/block.go]: err != io.EOF

File: disk/block.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+		buf:    make([]byte, blockSize),
+		h:      crc32.New(crcTable),
+	}
+}
+
+// ReadFrom implements io.ReaderFrom.
+// It reads from the reader to the underlying writer until EOF is hit and
+// returns the number of bytes read.
+// Any error except EOF encountered during the read is also returned.
+func (b *blockEncoder) ReadFrom(r io.Reader) (int64, error) {
```
Comment (xiang90) [inline disk/block.go]: EncodeFrom?

File: disk/block_test.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"math/rand"
+	"io/ioutil"
+	"bytes"
+	"testing"
+)
+
+func BenchmarkCRCEncoding(b *testing.B) {
+	c := b.N
```
Comment (xiang90) [inline disk/block_test.go]: can we just use b.N here?

**Comment by xiang90** [comment]
LGTM

File: disk/block.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	payload := b.buf[4:]
+	n, err := io.ReadAtLeast(r, payload, int(b.size - 4))
+	if err != nil && err != io.ErrUnexpectedEOF {
+		return 0, err
+	}
+
+	// fill zero
+	for i := n; i < len(payload); i++ {
+		payload[i] = 0
+	}
```
Comment (yutongp) [inline disk/block.go]: Is there a better way in Go to init slice to 0?@xiang90

Comment (xiang90) [inline disk/block.go]: this resets the slice. not init it.

Comment (yutongp) [inline disk/block.go]: It seems there is no good way to fill slice besides a loop

File: disk/block.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	// block size in bytes
+	// payload size = block size - 4
+	size int64
+	w    io.Writer
+	// optimization: global buf pool
+	buf    []byte // buf for payload
+	h      hash.Hash32
+}
+
+func NewBlockEncoder(blockSize int64, w io.Writer) *blockEncoder {
```
Comment (yutongp) [inline disk/block.go]: is blockEncoder a exported type or not?
If not, why NewBlockEncoder is a exported function?

**Pull request merged**